File: python/samples/getting_started/workflows/declarative/customer_support/ticketing_plugin.py
# Copyright (c) Microsoft. All rights reserved.

"""Ticketing plugin for CustomerSupport workflow."""

import uuid
from dataclasses import dataclass
from enum import Enum
from collections.abc import Callable

# ANSI color codes
MAGENTA = "\033[35m"
RESET = "\033[0m"


class TicketStatus(Enum):
    """Status of a support ticket."""

    OPEN = "open"
    IN_PROGRESS = "in_progress"
    RESOLVED = "resolved"
    CLOSED = "closed"


@dataclass
class TicketItem:
    """A support ticket."""

    id: str
    subject: str = ""
    description: str = ""
    notes: str = ""
    status: TicketStatus = TicketStatus.OPEN


class TicketingPlugin:
    """Mock ticketing plugin for customer support workflow."""

    def __init__(self) -> None:
        self._ticket_store: dict[str, TicketItem] = {}

    def _trace(self, function_name: str) -> None:
        print(f"\n{MAGENTA}FUNCTION: {function_name}{RESET}")

    def get_ticket(self, id: str) -> TicketItem | None:
        """Retrieve a ticket by identifier from Azure DevOps."""
        self._trace("get_ticket")
        return self._ticket_store.get(id)

    def create_ticket(self, subject: str, description: str, notes: str) -> str:
        """Create a ticket in Azure DevOps and return its identifier."""
        self._trace("create_ticket")
        ticket_id = uuid.uuid4().hex
        ticket = TicketItem(
            id=ticket_id,
            subject=subject,
            description=description,
            notes=notes,
        )
        self._ticket_store[ticket_id] = ticket
        return ticket_id

    def resolve_ticket(self, id: str, resolution_summary: str) -> None:
        """Resolve an existing ticket in Azure DevOps given its identifier."""
        self._trace("resolve_ticket")
        if ticket := self._ticket_store.get(id):
            ticket.status = TicketStatus.RESOLVED

    def send_notification(self, id: str, email: str, cc: str, body: str) -> None:
        """Send an email notification to escalate ticket engagement."""
        self._trace("send_notification")

    def get_functions(self) -> list[Callable[..., object]]:
        """Return all plugin functions for registration."""
        return [
            self.get_ticket,
            self.create_ticket,
            self.resolve_ticket,
            self.send_notification,
        ]

File: dotnet/samples/03-workflows/Declarative/CustomerSupport/TicketingPlugin.cs
// Copyright (c) Microsoft. All rights reserved.

using System.ComponentModel;

namespace Demo.Workflows.Declarative.CustomerSupport;

internal sealed class TicketingPlugin
{
    private readonly Dictionary<string, TicketItem> _ticketStore = [];

    [Description("Retrieve a ticket by identifier from Azure DevOps.")]
    public TicketItem? GetTicket(string id)
    {
        Trace(nameof(GetTicket));

        this._ticketStore.TryGetValue(id, out TicketItem? ticket);

        return ticket;
    }

    [Description("Create a ticket in Azure DevOps and return its identifier.")]
    public string CreateTicket(string subject, string description, string notes)
    {
        Trace(nameof(CreateTicket));

        TicketItem ticket = new()
        {
            Subject = subject,
            Description = description,
            Notes = notes,
            Id = Guid.NewGuid().ToString("N"),
        };

        this._ticketStore[ticket.Id] = ticket;

        return ticket.Id;
    }

    [Description("Resolve an existing ticket in Azure DevOps given its identifier.")]
    public void ResolveTicket(string id, string resolutionSummary)
    {
        Trace(nameof(ResolveTicket));

        if (this._ticketStore.TryGetValue(id, out TicketItem? ticket))
        {
            ticket.Status = TicketStatus.Resolved;
        }
    }

    [Description("Send an email notification to escalate ticket engagement.")]
    public void SendNotification(string id, string email, string cc, string body)
    {
        Trace(nameof(SendNotification));
    }

    private static void Trace(string functionName)
    {
        Console.ForegroundColor = ConsoleColor.DarkMagenta;
        try
        {
            Console.WriteLine($"\nFUNCTION: {functionName}");
        }
        finally
        {
            Console.ResetColor();
        }
    }

    public enum TicketStatus
    {
        Open,
        InProgress,
        Resolved,
        Closed,
    }

    public sealed class TicketItem
    {
        public TicketStatus Status { get; set; } = TicketStatus.Open;
        public string Subject { get; init; } = string.Empty;
        public string Id { get; init; } = string.Empty;
        public string Description { get; init; } = string.Empty;
        public string Notes { get; init; } = string.Empty;
    }
}
